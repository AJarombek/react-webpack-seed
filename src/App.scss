/**
 * Sass styling for the App component
 * @author Andrew Jarombek
 * @since 3/22/2018
 */

@font-face {
  font-family: Sylexiad;
  font-weight: normal;
  src: url('https://asset.jarombek.com/fonts/SylexiadSansThin.ttf') format('truetype');
}

@font-face {
  font-family: Sylexiad-Bold;
  font-weight: bold;
  src: url('https://asset.jarombek.com/fonts/SylexiadSansThin-Bold.ttf') format('truetype');
}

/**
 * A mixin that sets the font to sylexiad
 */
@mixin sylexiad {
  font: {
    family: 'Sylexiad', Helvetica, sans-serif !important;
    weight: normal;
    style: normal;
  }
  letter-spacing: 0.4px;
}

/**
 * A mixin that sets the font to sylexiad bold
 */
@mixin sylexiad-bold {
  font: {
    family: 'Sylexiad-Bold', Helvetica, sans-serif !important;
    weight: bold;
    style: normal;
  }
}

/**
 * A mixin that simplifies transition generation
 * @param $property - the property of the element that transitions
 * @param $result - the resulting style of the transition
 * @param $time - the time the transition takes to complete (optional)
 */
@mixin transition($property, $result, $time: 0.3s, $speed: ease) {
  transition: #{$property} $time #{$speed};

  // #{} syntax compiles the contents as a property name or selector
  &:hover {
    #{$property}: $result;
  }
}

body {
  background-color: #eee;
}

p {
  font-weight: bold;
  color: gray;
  @include sylexiad-bold;
}

.title-container {
  height: 160px;
  padding-top: 5px;
}

.title {
  color: #000;
  font-size: 28px;
  text-align: center;
  margin: 5% auto;
  @include transition('font-size', 28.6px, $speed: ease-in-out);
}

.sub-title {
  @include sylexiad;
  font-size: 24px;
  text-align: center;
}
