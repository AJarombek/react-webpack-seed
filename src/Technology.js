/**
 * A component that displays certain attributes of a technology
 * @author Andrew Jarombek
 * @since 3/22/2018
 */

import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';

import './Technology.scss';

const Technology = ({ name, release_date, picture }) => (
  <div className="technology">
    <figure>
      <img src={require(`${picture}`)} />
    </figure>
    <p className="tech-name">{name}</p>
    <p className="tech-release-date">{moment(release_date).format('MMMM Do, YYYY')}</p>
  </div>
);

Technology.propTypes = {
  name: PropTypes.string.isRequired,
  release_date: PropTypes.instanceOf(Date).isRequired,
  picture: PropTypes.string.isRequired,
};

export default Technology;
