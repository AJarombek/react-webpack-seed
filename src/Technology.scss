/**
 * @author Andrew Jarombek
 * @since 3/30/2018
 */

@import "App";

$darker-border: #aaa;

// The technology class is the grid container.  It has two rows and two columns
.technology {

  border-radius: 4px;
  border: #ccc 1px solid;
  @include transition("border-color", $darker-border);
  box-shadow: 0 1px 7px #ddd;
  background-color: #fff;

  // Define the grid layout
  display: grid;
  grid-template-columns: [image] 35% [text] 65% [end];
  grid-template-rows: [top] 50% [middle] 50% [bottom];

  // The image covers all rows in the first column
  figure {
    grid-column-start: image;
    grid-column-end: text;
    grid-row-start: top;
    grid-row-end: bottom;
    align-self: center;
    margin: auto;
  }

  img {
    height: 150px;
  }

  // The name covers the first row in the second column
  .tech-name {
    grid-column-start: text;
    grid-column-end: end;
    grid-row-start: top;
    grid-row-end: middle;
    align-self: end;
    font-size: 18px;
  }

  // The date covers the second row in the second column
  .tech-release-date {
    grid-column-start: text;
    grid-column-end: end;
    grid-row-start: middle;
    grid-row-end: bottom;
    align-self: start;
    font-size: 18px;
  }
}