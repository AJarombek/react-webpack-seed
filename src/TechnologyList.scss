/**
 * @author Andrew Jarombek
 * @since 3/30/3018
 */

// The technology-list div is the grid container for the technologies
.technology-list {

  // Creates a block level grid
  display: grid;

  // Define the columns in the grid.  In this case there is one column named 'main' that takes up 100%
  // of the grid containers width
  grid-template-columns: [main] 100%;

  // We don't have a defined number of rows in the grid.  Auto generate each one with a height of 200px
  grid-auto-rows: 200px;

  // Define the gap between each row in the grid as 20px and the gap between columns as 0
  grid-gap: 20px 0;

  justify-items: stretch;
}