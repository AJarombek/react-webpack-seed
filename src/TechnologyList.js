/**
 * A component that holds a list of technologies
 * @author Andrew Jarombek
 * @since 3/24/2018
 */

import React from 'react';
import Technology from './Technology';
import PropTypes from 'prop-types';

import './TechnologyList.scss';

const TechnologyList = ({ techList = [] }) => (
  <div className="technology-list">
    {techList.length === 0 ? (
      <p>No Data</p>
    ) : (
      techList.map((technology) => <Technology key={technology.id} {...technology} />)
    )}
  </div>
);

TechnologyList.propTypes = {
  techList: PropTypes.array,
};

export default TechnologyList;
