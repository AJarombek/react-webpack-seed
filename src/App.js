/**
 * Create a root component for the react application
 * @author Andrew Jarombek
 * @since 3/21/2018
 */

import React from 'react';
import {hot} from 'react-hot-loader';
import TechnologyList from './TechnologyList';
import moment from 'moment'

import './App.scss';

class App extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            technologies: [
                {
                    id: "375e6c2b-d799-47ef-848d-84b42bef7a29",
                    name: "JavaScript",
                    picture: "./assets/js.png",
                    release_date: moment('1995-12-04')
                },
                {
                    id: "75bb50f5-9067-4a30-964f-a3f0d42b5ca4",
                    name: "Webpack",
                    picture: "./assets/webpack.png",
                    release_date: moment('2012-03-10')
                },
                {
                    id: "ed13c757-27cd-4cc1-b4ac-de449ee9cbae",
                    name: "React",
                    picture: "./assets/react.png",
                    release_date: moment('2013-03-01')
                },
                {
                    id: "d9f08c13-f59f-45bb-b8e9-0bd0114b0adf",
                    name: "Sass",
                    picture: "./assets/sass.png",
                    release_date: moment('2006-11-28')
                }
            ]
        }
    }

    render() {
        const {technologies} = this.state;
        return (
            <div>
                <div className="title-container">
                    <p className="title">React Webpack Seed</p>
                </div>
                <p className="sub-title">Technologies used</p>
                <TechnologyList techList={technologies} />
            </div>
        );
    }
}

// Enable Hot Module Replacement on this component
export default hot(module)(App);