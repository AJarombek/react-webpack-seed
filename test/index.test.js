/**
 * Test the root App component with Jest and Enzyme
 * @author Andrew Jarombek
 * @since 3/21/2018
 */

import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import App from '../src/App';
import TechnologyList from '../src/TechnologyList';
import Technology from '../src/Technology';

const techList = [
  {
    id: 'ed13c757-27cd-4cc1-b4ac-de449ee9cbae',
    name: 'React',
    picture: './assets/react.png',
    release_date: new Date('2013-03-01'),
  },
  {
    id: 'd9f08c13-f59f-45bb-b8e9-0bd0114b0adf',
    name: 'Sass',
    picture: './assets/sass.png',
    release_date: new Date('2006-11-28'),
  },
];

const app = shallow(<App />);
const technologyList = shallow(<TechnologyList techList={techList} />);
const techListNoProps = shallow(<TechnologyList />);

const technology = shallow(
  <Technology name={techList[0].name} picture={techList[0].picture} release_date={techList[0].release_date} />,
);

// We must mock the format() function for the moment() call
// The jest.mock() function expects one argument - a function that returns the mock
jest.mock('moment', () => () => ({ format: () => 'March 1st, 2013' }));

test('App has Title', () => {
  expect(app.find('.title')).toHaveLength(1);
});

test('App has Sub-Title', () => {
  expect(app.find('.sub-title')).toHaveLength(1);
});

test('Technology List has two technologies', () => {
  expect(technologyList.find('.technology-list').length).toBe(1);
});

// Snapshot test - every time a test is run Jest will compare against the previous snapshot
test('Technology List with No Props', () => {
  expect(toJSON(techListNoProps)).toMatchSnapshot();
});

test('Technology name is React', () => {
  expect(technology.find('.tech-name').text()).toEqual('React');
});

test('Technology release date is march 2013', () => {
  expect(technology.find('.tech-release-date').text()).toEqual('March 1st, 2013');
});
