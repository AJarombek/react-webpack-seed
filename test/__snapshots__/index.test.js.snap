// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Technology List with No Props 1`] = `
<div
  className="technology-list"
>
  <p>
    No Data
  </p>
</div>
`;
